File: python/leecode/algo/algo1847.py
class Solution:
    def closestRoom(self, rooms, queries) :
        rooms.sort(key = lambda p: p[1])
        count = len(queries)
        ret = [-1] * count
        roomids = []
        
        from bisect import bisect_right
        def put(roomid):
            inx = bisect_right(roomids, roomid)
            roomids.insert(inx, roomid)
            
        for ri, query in sorted(enumerate(queries), key = lambda p:p[1][1], reverse = True):
            while rooms and rooms[-1][1] >= query[1]:
                roomid, _ = rooms.pop(-1)
                put(roomid)
            if roomids:
                inx = bisect_right(roomids, query[0])
                if inx == 0:
                    ret[ri] = roomids[0]
                elif inx == len(roomids):
                    ret[ri] = roomids[inx - 1]
                else:
                    ret[ri] = roomids[inx - 1] if (roomids[inx-1] + roomids[inx]) // 2 >= query[0] else roomids[inx]
        return ret

if __name__ == "__main__":
    sol = Solution()
    # rooms = [[2,2],[1,2],[3,2]]
    # queries = [[3,1],[3,3],[5,2]]
    # rooms = [[1,4],[2,3],[3,5],[4,1],[5,2]]
    # queries = [[2,3],[2,4],[2,5]]
    rooms = [[23,22],[6,20],[15,6],[22,19],[2,10],[21,4],[10,18],[16,1],[12,7],[5,22]]
    queries = [[12,5],[15,15],[21,6],[15,1],[23,4],[15,11],[1,24],[3,19],[25,8],[18,6]]
    print(sol.closestRoom(rooms, queries))        

File: cplus/leetcode/202508/algo1847.cpp
#include <bits/stdc++.h>
using namespace std;

// 常用数据结构 - 专题 - 离线算法
class Solution {
public:
    vector<int> closestRoom(vector<vector<int>>& rooms, vector<vector<int>>& queries) {
        // 复习 - 自行解答
        sort(rooms.begin(), rooms.end(), [](const auto& a, const auto& b) {
             return a[1] > b[1];
        });

        int l = 0, n = rooms.size(), m = queries.size();
        int idx[m];
        iota(idx, idx+m, 0);
        sort(idx, idx + m, [&](int i, int j) {
            return queries[i][1] > queries[j][1];
        });

        set<int> r_ids; vector<int> ans(m);
        for (int i=0; i<m; i++) {
            int id = idx[i];
            while (l < n && rooms[l][1] >= queries[id][1]) {
                r_ids.insert(rooms[l++][0]);
            }
            if (r_ids.empty()) {
                ans[id] = -1;
            } else {
                auto it = r_ids.lower_bound(queries[id][0]);
                if (it == r_ids.end()) {
                    ans[id]  = *prev(it);
                } else if (it == r_ids.begin()) {
                    ans[id]  = *it;
                } else {
                    ans[id] = *it - queries[id][0] < queries[id][0] - *prev(it)? *it : *prev(it);
                }
            }
        }
        return ans; 
    }
};